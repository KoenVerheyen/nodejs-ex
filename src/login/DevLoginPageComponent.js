// @flow
import React from 'react'
import {Panel} from "nes-core-frontend/lib/components/core/Panel";
import {Select, Button} from 'antd'
import IdentityFormSelect from "./IdentityFormSelect";

const Option = Select.Option
const PERSON = 'person'
const FUNCTION = 'function'
const PROFILE = 'profile'
const ID = 'id'
const LABEL_LENGTH = 100

function filterByChildPropId(array, prop, value = undefined) {
    if (value) {
        return array.filter((elem) => elem[prop][ID] === value)
    }
    return array.slice() // a copy of the array
}

function filterUniqueById(array) {
    let flags = {}
    return array.filter(function (entry) {
        if (flags[entry[ID]]) {
            return false
        }
        flags[entry[ID]] = true
        return true
    })
}

function reduceByProp(array, prop) {
    return array.reduce(function (prev, curr) {
        return [...prev, curr[prop]]
    }, [])
}

type Props = {
    +token: string,
    +identities: Array<Object>,
    +loading: boolean,
    +onSelectToken: Function,
    +postSelectIdentity: Function,
}
type Identity = {
    person: number,
    function?: number,
    profile?: number,
}
type State = {
    identity: Identity
}


export default class DevLoginPageComponent extends React.Component<Props, State> {

    constructor(props: Props) {
        super(props)
        this.state = {
            identity: {
                person: 0,
            }
        }
    }

    componentWillReceiveProps(nextProps : Props) {
        if (nextProps.identities.length > 0 && !nextProps.loading) {
            this.setState({
                identity: {
                    person : Number(nextProps.identities[0].person.id),
                    function: nextProps.identities[0].function.id,
                    profile: nextProps.identities[0].profile.id,
                }

            })
        }
    }

    onPersonChange(value : string) {
        const firstIdentity = filterByChildPropId(this.props.identities, PERSON, Number(value))[0]
        this.setState({
            identity: {
                person: Number(value),
                function: firstIdentity.function.id,
                profile: firstIdentity.profile.id
            }
        })
    }

    onFunctionChange(value : string) {
        const firstIdentity = filterByChildPropId(this.props.identities, FUNCTION, Number(value))[0]
        this.setState({
            identity: {
                person: firstIdentity.person.id,
                function: Number(value),
                profile: firstIdentity.profile.id
            }
        })
    }

    onProfileChange(value: string) {
        let identity = {
            ...this.state.identity,
            profile: Number(value),
        }
        this.setState({identity})
    }

    onSubmitIdentity() {
        this.props.postSelectIdentity(this.props.token, this.state.identity)
    }

    getPersons() {
        const identities = this.props.identities.slice()
        return filterUniqueById(reduceByProp(identities, PERSON))
    }

    getFunctions() {
        const identities = this.props.identities.slice()
        const id = this.state.identity.person || undefined
        return filterUniqueById(reduceByProp(filterByChildPropId(identities, PERSON, id), FUNCTION))
    }

    getProfiles() {
        const identities = this.props.identities.slice()
        const id = this.state.identity.function || undefined
        return filterUniqueById(reduceByProp(filterByChildPropId(identities, FUNCTION, id), PROFILE))
    }


    render() {
        return (<Panel style={{ display: 'block'}}>
            <div style={{ display: 'flex'}}>
            <div style={{width: LABEL_LENGTH}}>{'Token: '}</div>
            <Select value={this.props.token} onChange={this.props.onSelectToken} style={{width: 200}}>
                <Option value='lrc'>Local Rad Coordinator</Option>
                <Option value='nrc'>National Rad Coordinator</Option>
                <Option value='rad'>Rad Team</Option>
            </Select></div>
            {this.props.identities.length > 0 ? (<div style={{ display: 'flex'}}>
                <div style={{width: LABEL_LENGTH}}>Person: </div>
                <IdentityFormSelect
                    name={PERSON}
                    value={this.state.identity.person}
                    onSelectChange={this.onPersonChange.bind(this)}
                    elements={this.getPersons()}
                    getName={(elem) => elem.firstName && elem.lastName ? elem.firstName + ' ' + elem.lastName : 'Unknown' }
                    style={{width: 200}}
                /></div>):undefined}
                {this.props.identities.length > 0 ? (<div style={{ display: 'flex'}}>
                <div style={{width: LABEL_LENGTH}}>Function: </div>
                <IdentityFormSelect
                    name={FUNCTION}
                    value={this.state.identity.function}
                    onSelectChange={this.onFunctionChange.bind(this)}
                    elements={this.getFunctions()}
                    getName={(elem) => elem.name? elem.name: 'unknown'}
                    style={{width: 200}}
                /></div>): undefined}
            {this.props.identities.length > 0 ? (<div style={{ display: 'flex'}}>
                <div style={{width: LABEL_LENGTH}}>Profile: </div>
                <IdentityFormSelect
                    name={PROFILE}
                    value={this.state.identity.profile}
                    onSelectChange={this.onProfileChange.bind(this)}
                    elements={this.getProfiles()}
                    getName={(elem) => elem.name ? elem.name: 'unknown'}
                    style={{width: 200}}
                />
            </div>) : undefined}
            {this.props.identities.length > 0 ? (<div style={{ display: 'flex'}}>
                <Button onClick={this.onSubmitIdentity.bind(this)} loading={this.props.loading}>Login</Button>
            </div>) : undefined}

        </Panel>)
    }
}
