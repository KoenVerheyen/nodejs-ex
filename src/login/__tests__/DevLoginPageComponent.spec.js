import React from 'react'
import {shallow} from 'enzyme'
import DevLoginPageComponent from "../DevLoginPageComponent";


describe('DevLoginComponent', () => {

    test('initial', () => {
        let wrapper = shallow(<DevLoginPageComponent identities={[]}/>)

        expect(wrapper).toMatchSnapshot()
        expect(wrapper.length).toEqual(1)
        let optionList = wrapper.find('Option')
        expect(optionList.length).toEqual(3)
        expect(optionList.get(0).props.value).toEqual('lrc')
        expect(optionList.get(1).props.value).toEqual('nrc')
        expect(optionList.get(2).props.value).toEqual('rad')
    })


    test('selected token', () => {
        let identities = [
            {
                person: {
                    id: 1,
                    lastName: 'Doe',
                    firstName: 'John'
                },
                function: {
                    id: 1,
                    name: 'functionLF'
                },
                profile: {
                    id: 1,
                    name: 'local rad coordinator'
                }
            },
            {
                person: {
                    id: 1,
                    lastName: 'Doe',
                    firstName: 'John'
                },
                function: {
                    id: 2,
                    name: 'functionEB'
                },
                profile: {
                    id: 1,
                    name: 'local rad coordinator'
                }
            }
        ]
        let submitIdentity = jest.fn()
        let wrapper = shallow(<DevLoginPageComponent token='lrc' identities={identities} postSelectIdentity={submitIdentity}/>)

        expect(wrapper).toMatchSnapshot()
        expect(wrapper.length).toEqual(1)
        let select = wrapper.find('Select')
        expect(select.length).toEqual(1)
        expect(select.props().value).toEqual('lrc')

        let person = wrapper.find('IdentityFormSelect[name="person"]')
        expect(person.length).toEqual(1)
        expect(person.props().getName(identities[0].person)).toEqual('John Doe')
        expect(wrapper.state().identity.person).not.toEqual(1)
        person.props().onSelectChange(1)
        expect(wrapper.state().identity.person).toEqual(1)

        let functionSelect = wrapper.find('IdentityFormSelect[name="function"]')
        expect(functionSelect.length).toEqual(1)
        expect(functionSelect.props().getName(identities[0].function)).toEqual('functionLF')
        functionSelect.props().onSelectChange(1)
        expect(wrapper.state().identity.function).toEqual(1)

        let profile = wrapper.find('IdentityFormSelect[name="profile"]')
        expect(profile.length).toEqual(1)
        expect(profile.props().getName(identities[0].profile)).toEqual('local rad coordinator')
        profile.props().onSelectChange(1)
        expect(wrapper.state().identity.profile).toEqual(1)


        let loginButton = wrapper.find('Button')
        expect(loginButton.length).toEqual(1)
        loginButton.simulate('click')
        expect(submitIdentity).toHaveBeenCalledTimes(1)
    })



    test('selected token loading', () => {
        let identities = [
            {
                person: {
                    id: 1,
                    lastName: 'Doe',
                    firstName: 'John'
                },
                function: {
                    id: 1,
                    name: 'functionLF'
                },
                profile: {
                    id: 1,
                    name: 'local rad coordinator'
                }
            }
        ]
        let wrapper = shallow(<DevLoginPageComponent token='lrc' identities={identities} loading={true}/>)

        expect(wrapper).toMatchSnapshot()

        let loginButton = wrapper.find('Button')
        expect(loginButton.length).toEqual(1)
        expect(loginButton.props().loading).toEqual(true)
    })

})