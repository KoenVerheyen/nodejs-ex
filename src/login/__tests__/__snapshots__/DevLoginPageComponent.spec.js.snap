// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DevLoginComponent initial 1`] = `
<Panel
  style={
    Object {
      "display": "block",
    }
  }
>
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <div
      style={
        Object {
          "width": 100,
        }
      }
    >
      Token: 
    </div>
    <Select
      choiceTransitionName="zoom"
      prefixCls="ant-select"
      showSearch={false}
      style={
        Object {
          "width": 200,
        }
      }
      transitionName="slide-up"
    >
      <Option
        value="lrc"
      >
        Local Rad Coordinator
      </Option>
      <Option
        value="nrc"
      >
        National Rad Coordinator
      </Option>
      <Option
        value="rad"
      >
        Rad Team
      </Option>
    </Select>
  </div>
</Panel>
`;

exports[`DevLoginComponent selected token 1`] = `
<Panel
  style={
    Object {
      "display": "block",
    }
  }
>
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <div
      style={
        Object {
          "width": 100,
        }
      }
    >
      Token: 
    </div>
    <Select
      choiceTransitionName="zoom"
      prefixCls="ant-select"
      showSearch={false}
      style={
        Object {
          "width": 200,
        }
      }
      transitionName="slide-up"
      value="lrc"
    >
      <Option
        value="lrc"
      >
        Local Rad Coordinator
      </Option>
      <Option
        value="nrc"
      >
        National Rad Coordinator
      </Option>
      <Option
        value="rad"
      >
        Rad Team
      </Option>
    </Select>
  </div>
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <div
      style={
        Object {
          "width": 100,
        }
      }
    >
      Person: 
    </div>
    <IdentityFormSelect
      elements={
        Array [
          Object {
            "firstName": "John",
            "id": 1,
            "lastName": "Doe",
          },
        ]
      }
      getName={[Function]}
      name="person"
      onSelectChange={[Function]}
      style={
        Object {
          "width": 200,
        }
      }
      value={0}
    />
  </div>
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <div
      style={
        Object {
          "width": 100,
        }
      }
    >
      Function: 
    </div>
    <IdentityFormSelect
      elements={
        Array [
          Object {
            "id": 1,
            "name": "functionLF",
          },
          Object {
            "id": 2,
            "name": "functionEB",
          },
        ]
      }
      getName={[Function]}
      name="function"
      onSelectChange={[Function]}
      style={
        Object {
          "width": 200,
        }
      }
    />
  </div>
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <div
      style={
        Object {
          "width": 100,
        }
      }
    >
      Profile: 
    </div>
    <IdentityFormSelect
      elements={
        Array [
          Object {
            "id": 1,
            "name": "local rad coordinator",
          },
        ]
      }
      getName={[Function]}
      name="profile"
      onSelectChange={[Function]}
      style={
        Object {
          "width": 200,
        }
      }
    />
  </div>
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <Button
      ghost={false}
      loading={false}
      onClick={[Function]}
      prefixCls="ant-btn"
    >
      Login
    </Button>
  </div>
</Panel>
`;

exports[`DevLoginComponent selected token loading 1`] = `
<Panel
  style={
    Object {
      "display": "block",
    }
  }
>
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <div
      style={
        Object {
          "width": 100,
        }
      }
    >
      Token: 
    </div>
    <Select
      choiceTransitionName="zoom"
      prefixCls="ant-select"
      showSearch={false}
      style={
        Object {
          "width": 200,
        }
      }
      transitionName="slide-up"
      value="lrc"
    >
      <Option
        value="lrc"
      >
        Local Rad Coordinator
      </Option>
      <Option
        value="nrc"
      >
        National Rad Coordinator
      </Option>
      <Option
        value="rad"
      >
        Rad Team
      </Option>
    </Select>
  </div>
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <div
      style={
        Object {
          "width": 100,
        }
      }
    >
      Person: 
    </div>
    <IdentityFormSelect
      elements={
        Array [
          Object {
            "firstName": "John",
            "id": 1,
            "lastName": "Doe",
          },
        ]
      }
      getName={[Function]}
      name="person"
      onSelectChange={[Function]}
      style={
        Object {
          "width": 200,
        }
      }
      value={0}
    />
  </div>
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <div
      style={
        Object {
          "width": 100,
        }
      }
    >
      Function: 
    </div>
    <IdentityFormSelect
      elements={
        Array [
          Object {
            "id": 1,
            "name": "functionLF",
          },
        ]
      }
      getName={[Function]}
      name="function"
      onSelectChange={[Function]}
      style={
        Object {
          "width": 200,
        }
      }
    />
  </div>
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <div
      style={
        Object {
          "width": 100,
        }
      }
    >
      Profile: 
    </div>
    <IdentityFormSelect
      elements={
        Array [
          Object {
            "id": 1,
            "name": "local rad coordinator",
          },
        ]
      }
      getName={[Function]}
      name="profile"
      onSelectChange={[Function]}
      style={
        Object {
          "width": 200,
        }
      }
    />
  </div>
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <Button
      ghost={false}
      loading={true}
      onClick={[Function]}
      prefixCls="ant-btn"
    >
      Login
    </Button>
  </div>
</Panel>
`;
